whole-tale/girder_wholetale
# Allow to import an external dataset as a Tale

### Rationale
Default "Analyze in WT" workflow imports incoming dataset as an immutable, readonly data. In some scenarios (e.g. a Binder compatible collection of files published to Dataverse whole-tale/whole-tale#66) this behavior is undesirable. Data should be copied to the Workspace instead to allow user to modify the environment and/or the code.

### Approach
1. Add a parameter `asTale` to `POST /tale/import` to indicate whether the dataset defined by `url` parameter should be imported as involatile data or copied to the workspace.
2. Importing as Tale is implemented as follows:
   1. Register the dataset defined by the `url` in the Catalog
   2. Create a `Session` with the content of registered folder from 1.
   3. Create an empty `Tale`.
   4. Copy data from WtDms (defined by the Session from step 2.) to the workspace (created in step 3.)

### How to test?

Make sure you pull `wholetale/girder` before testing this. I added new dependency: https://github.com/whole-tale/girder/commit/2ad286c4fe186fbff3fc63b85c0e67d0fc2e8480

1. (optionally): Verify that
  ```
  curl -IL "https://girder.local.wholetale.org/api/v1/integration/dataverse?fileId=30&siteUrl=https://dev2.dataverse.org"
  ```
  returns uri to entire dataset instead of a single file. Compare with the output of:
  ```
  curl -IL "https://girder.local.wholetale.org/api/v1/integration/dataverse?fileId=30&siteUrl=https://dev2.dataverse.org&fullDataset=false"
  ```
  which is a previous default behaviour.
1. Add "https://dev2.dataverse.org" to `Extra Dataverse Hosts` (Girder UI > Admin console > Plugins > Wholetale [link](https://girder.local.wholetale.org/#plugins/wholetale/config)).
1. Navigate to https://dev2.dataverse.org/file.xhtml?fileId=30&version=1.0, select "Explore > girder.local Whole Tale".
1. Confirm that Input data contains: `Data Source: https://dev2.dataverse.org/dataset.xhtml?persistentId=doi:10.5072/FK2/NYNHAM`
2. Choose image, and hit "Launch New Tale". **NOTE:** I'm not actually spawning a Tale right now, so after job finishes, the dashboard returns to the default state of compose
3. Navigate to Browse.
4. Confirm that "Dataverse IRC Metrics" Tale is there and run it.
5. Confirm that workspace contains a bunch of files. 

### TODO:
1. Actually launch Tale in step 2.
1. Add tests
1. This particular example requires latest repo2docker_wholetale, that I haven't fully tested yet. To enable it, you need add new custom image with `buildpack=="RBuildPack"`
